src/shork/pool.gleam

```diff
@@ -0,0 +1,231 @@
+import gleam/erlang/process.{type Subject}
+import gleam/list
+import gleam/option.{type Option, None, Some}
+import gleam/otp/actor.{type Next}
+import gleam/result
+import shork
+
+pub opaque type PoolConfig {
+  PoolConfig(
+    connection_config: shork.Config,
+    pool_size: Int,
+    max_age: Int,
+    tick_interval: Int,
+    connection_timeout: Int,
+  )
+}
+
+type PoolConnection {
+  PoolConnection(
+    connection: Option(shork.Connection),
+    in_use: Bool,
+    is_outside_of_pool: Bool,
+    age: Int,
+  )
+}
+
+type PoolState {
+  PoolState(config: PoolConfig, connections: List(PoolConnection))
+}
+
+pub opaque type ConnectionPool {
+  ConnectionPool(actor: Subject(PoolActorMessage), config: PoolConfig)
+}
+
+pub fn default_config(connection_config: shork.Config) -> PoolConfig {
+  PoolConfig(
+    connection_config: connection_config,
+    pool_size: 10,
+    max_age: 20_000,
+    tick_interval: 250,
+    connection_timeout: 1000,
+  )
+}
+
+pub fn pool_size(config: PoolConfig, pool_size: Int) -> PoolConfig {
+  PoolConfig(..config, pool_size: pool_size)
+}
+
+pub fn max_age(config: PoolConfig, max_age: Int) -> PoolConfig {
+  PoolConfig(..config, max_age: max_age)
+}
+
+pub fn create(config: PoolConfig) -> Result(ConnectionPool, actor.StartError) {
+  let initial_state =
+    PoolState(
+      config: config,
+      connections: list.range(1, config.pool_size)
+        |> list.map(fn(_) {
+          PoolConnection(
+            connection: None,
+            in_use: False,
+            age: 0,
+            is_outside_of_pool: False,
+          )
+        }),
+    )
+
+  use pool_actor <- result.try(actor.start(initial_state, pool_actor_handler))
+  actor.send(pool_actor, Tick(pool_actor))
+
+  Ok(ConnectionPool(pool_actor, config))
+}
+
+type ConnectionActorMessage {
+  ConnectionEstablished(connection: PoolConnection)
+}
+
+type PoolActorMessage {
+  Tick(Subject(PoolActorMessage))
+  Connect(Subject(ConnectionActorMessage))
+  Disconnect(PoolConnection)
+}
+
+pub fn connect(pool: ConnectionPool) -> Result(shork.Connection, Nil) {
+  process.try_call(
+    pool.actor,
+    fn(connection_actor) { Connect(connection_actor) },
+    pool.config.connection_timeout,
+  )
+  |> result.map(fn(response: ConnectionActorMessage) {
+    let pool_connection = response.connection
+    let assert Some(shork_connection) = pool_connection.connection
```

Comment:
What happens this is none?

Reply:
Good catch! I guess my implicit assumption was that find_unused_and_mark_as_used() will not return connections that are not established yet...

Reply:
fixed.